sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala

```diff
@@ -543,6 +543,12 @@ object ScalaReflection extends ScalaReflection {
             throw new UnsupportedOperationException(s"`$fieldName` is a reserved keyword and " +
               "cannot be used as field name\n" + walkedTypePath)
           }
+          // It is ok to use head here because a variable name needs to be non empty
+          if (fieldName.head.isDigit) {
```

Comment:
Rather than special case this, because there would be other ways this can fail, is it possible to improve the error closer to the source? or catch the right error around here and rethrow with clarification?

Reply:
I am no expert, but I would say this is the best/only way to catch it. Because when you generate the code it fails because Java variables cannot start with digits. So it is needed to be caught before the code is generated, because after we can't, I think.

Reply:
This doesn't catch all cases where a field name causes a problem though. I think this is too ad hoc

Reply:
There is already one case for when the name is a Java keyword, this is the case when the name starts with a number, also possible in Scala but not in Java. I think it is better to "customize" exceptions explaining why the input cannot be like that, than having the exception with 0 information and the user without knowing what to do.